### PR TITLE
Runtime cl device selection

### PR DESCRIPTION
**Pull request by t4c1**
## Summary
Adds an option to select the OpenCL device to use at runtime. By default we stil use the device selected by macros.

Implementation of `kernel_cl` is also simplified.

## Tests
Added a test that checks everything works when switching devices. However that test truly checks anything only if it is running on a computer with multiple devices installed.

## Side Effects
Non-generated OpenCL kernels are compiled when used for the first time. This makes them consistent with how kernel generator works.

## Release notes
Added an option to select the OpenCL device to use at runtime.

## Checklist

- [ ] Math issue #(issue number)

- [x] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: test/unit/math/opencl/opencl_context_test.cpp [hunk 1]
```diff
@@ -74,4 +75,38 @@ TEST(opencl_context, compile_kernel_rawcode) {
...
+               std::system_error);
+  EXPECT_THROW(stan::math::opencl_context.select_device(0, -1),
+               std::system_error);
+  EXPECT_THROW(stan::math::opencl_context.select_device(99999, 0),
+               std::system_error);
+  EXPECT_THROW(stan::math::opencl_context.select_device(0, 99999),
+               std::system_error);
+}
+
+// this test only checks anything if there are multiple devices installed
```
Comment (rok-cesnovar) [inline test/unit/math/opencl/opencl_context_test.cpp:91]:
```suggestion
// Checks that select_device() works for all devices found on the system. If there are multiple devices, this also tests that select_device() calls work after another device was already in use.
```

**Review by rok-cesnovar — CHANGES_REQUESTED** [review]
Looks good. I have one minor question and a suggestion for a more informative comment.

File: stan/math/opencl/kernel_cl.hpp [hunk 2]
```diff
@@ -214,9 +214,15 @@ inline const std::vector<cl::Event> select_events(
  */
 inline auto compile_kernel(const char* name,
                            const std::vector<std::string>& sources,
-                           std::map<std::string, int>& options) {
+                           const std::map<std::string, int>& options) {
+  auto base_opts = opencl_context.base_opts();
+  for (auto& it : options) {
+    if (base_opts[it.first] > it.second) {
```
[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/opencl/kernel_cl.hpp:220]: I think we had this discussion once. I think at the end you convinced me that > is the correct thing to use, although I don't remember the arguments.

[the remark this answers is not in the document]
Comment (rok-cesnovar) [reply stan/math/opencl/kernel_cl.hpp:220]: I think I now remember that the problem was if a device has a lower limit, say max group size is 10 and the compile_kernel calls with a higher limit that could be a problem. There is probably a better way of doing all this but that is a separate issue/PR.

**Review by rok-cesnovar — APPROVED** [review]

**Comment by t4c1** [comment]
The force push just removes all the debbuging code.

**Review by rok-cesnovar — APPROVED** [review]
LGTM.

Tested with the cmdstan branch and this seems to work great. Will open the cmdnstan PR  once this propagates upstream.

**Pull request merged**